src/utils/service.py

```diff
@@ -0,0 +1,203 @@
+"""Utility module containing logic of creation and deletion of managed Kyuubi K8s service."""
+
+import enum
+import functools
+import socket
+import typing
+
+import lightkube
+from ops import Model
+
+from constants import JDBC_PORT
+from utils.logging import WithLogging
+
+
+class _ServiceType(enum.Enum):
+    """Supported K8s service types."""
+
+    CLUSTER_IP = "ClusterIP"
+    NODE_PORT = "NodePort"
+    LOAD_BALANCER = "LoadBalancer"
+
+
+class ServiceUtil(WithLogging):
+    """Utility class containing logic of creation and deletion of managed Kyuubi K8s service."""
+
+    def __init__(self, model: Model):
+        self.model = model
+        self.app_name = model.app.name
+        self.service_name = f"{self.app_name}-service"
+        self.namespace = model.name
+        self.lightkube_agent = lightkube.Client()
+
+    @property
+    def model_service_domain(self) -> str:
+        """K8s service domain for Juju model."""
+        # Example: "kyuubi-k8s-0.kyuubi-k8s-endpoints.my-model.svc.cluster.local"
+        fqdn = socket.getfqdn()
+        # Example: "kyuubi-k8s-0.kyuubi-k8s-endpoints."
+        prefix = f"{self.model.unit.name.replace('/', '-')}.{self.app_name}-endpoints."
+        assert fqdn.startswith(f"{prefix}{self.namespace}.")
+        # Example: my-model.svc.cluster.local
+        return fqdn.removeprefix(prefix)
+
+    @property
+    def _host(self) -> str:
+        """K8s service hostname for Kyuubi."""
+        # Example: kyuubi-k8s-service.my-model.svc.cluster.local
+        return f"{self.service_name}.{self.model_service_domain}"
+
+    def _get_node_host(self) -> set[str]:
+        """Return the node ports of nodes where units of this app are scheduled."""
+
+        def _get_node_address(node) -> str:
```

Comment:
**suggestion**: we could inline this function content, since we do not pass it to a `map()` or something like that

Reply:
Nested function used here to break from nested loop. An inline solution would need:

```python
result = None

for ...1stloop...:
    out = False
    for ...2ndloop...:
        if something:
            result = something
            out = True
            break
        if out:
            break
```

which doesn't please my eyes much.

Instead, the following looks pretty:

```python
def get_something():
    for ...1stloop...:
        for ...2ndloop...:
            if something:
                return something
                
 result = get_something()
```